### Import Libraries

In [16]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.metrics import confusion_matrix, classification_report
import seaborn as sns
import matplotlib.pyplot as plt

### Import Dataframe

In [14]:
import pandas as pd

df = pd.read_csv("C:/Users/Lenovo/Downloads/Machine Learning Assignment/Machine Learning Assignment/CrimeData_from_2020_to_Present.csv", low_memory=False)
df_model

Date Rptd                DATE OCC  TIME OCC  \
0        211507896          11/4/2021 0:00          7/11/2020 0:00       845   
1        201516622  10/21/2020 12:00:00 AM  10/18/2020 12:00:00 AM      1845   
2        240913563         10/12/2024 0:00  10/30/2020 12:00:00 AM      1240   
3        210704711  12/24/2020 12:00:00 AM  12/24/2020 12:00:00 AM      1310   
4        201418201          3/10/2020 0:00  09/29/2020 12:00:00 AM      1830   
...            ...                     ...                     ...       ...   
1004986  252104112           2/2/2025 0:00           2/2/2025 0:00       130   
1004987  250404100  02/18/2025 12:00:00 AM  02/18/2025 12:00:00 AM      1000   
1004988  251304095  01/31/2025 12:00:00 AM  01/30/2025 12:00:00 AM      1554   
1004989  251704066  01/17/2025 12:00:00 AM  01/17/2025 12:00:00 AM      1600   
1004990  251904210  03/25/2025 12:00:00 AM  03/25/2025 12:00:00 AM      1235   

         AREA    AREA NAME  Rpt Dist No  Part 1-2  Crm Cd  \
0          15  N Hollywood         1502         2     354   
1          15  N Hollywood         1521         1     230   
2           9     Van Nuys          933         2     354   
3           7     Wilshire          782         1     331   
4          14      Pacific         1454         1     420   
...       ...          ...          ...       ...     ...   
1004986    21      Topanga         2103         2     946   
1004987     4   Hollenbeck          479         2     237   
1004988    13       Newton         1372         2     850   
1004989    17   Devonshire         1774         2     624   
1004990    19      Mission         1944         2     850   

                                               Crm Cd Desc  ... Status  \
0                                        THEFT OF IDENTITY  ...     IC   
1           ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT  ...     IC   
2                                        THEFT OF IDENTITY  ...     IC   
3        THEFT FROM MOTOR VEHICLE - GRAND ($950.01 AND ...  ...     IC   
4          THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)  ...     IC   
...                                                    ...  ...    ...   
1004986                          OTHER MISCELLANEOUS CRIME  ...     IC   
1004987                     CHILD NEGLECT (SEE 300 W.I.C.)  ...     IC   
1004988                                  INDECENT EXPOSURE  ...     IC   
1004989                           BATTERY - SIMPLE ASSAULT  ...     IC   
1004990                                  INDECENT EXPOSURE  ...     IC   

         Status Desc Crm Cd 1 Crm Cd 2  Crm Cd 3 Crm Cd 4  \
0        Invest Cont    354.0      NaN       NaN      NaN   
1        Invest Cont    230.0      NaN       NaN      NaN   
2        Invest Cont    354.0      NaN       NaN      NaN   
3        Invest Cont    331.0      NaN       NaN      NaN   
4        Invest Cont    420.0      NaN       NaN      NaN   
...              ...      ...      ...       ...      ...   
1004986  Invest Cont    946.0      NaN       NaN      NaN   
1004987  Invest Cont    237.0      NaN       NaN      NaN   
1004988  Invest Cont    850.0      NaN       NaN      NaN   
1004989  Invest Cont    624.0      NaN       NaN      NaN   
1004990  Invest Cont    850.0      NaN       NaN      NaN   

                                         LOCATION Cross Street      LAT  \
0         7800    BEEMAN                       AV          NaN  34.2124   
1                 ATOLL                        AV     N  GAULT  34.1993   
2        14600    SYLVAN                       ST          NaN  34.1847   
3         6000    COMEY                        AV          NaN  34.0339   
4                         4700    LA VILLA MARINA          NaN  33.9813   
...                                           ...          ...      ...   
1004986  22100    ROSCOE                       BL          NaN  34.2259   
1004987   3500    PERCY                        ST          NaN  34.0277   
1004988    300 E  53RD                    

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1004991 entries, 0 to 1004990
Data columns (total 28 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0                   1004991 non-null  int64  
 1   Date Rptd       1004991 non-null  object 
 2   DATE OCC        1004991 non-null  object 
 3   TIME OCC        1004991 non-null  int64  
 4   AREA            1004991 non-null  int64  
 5   AREA NAME       1004991 non-null  object 
 6   Rpt Dist No     1004991 non-null  int64  
 7   Part 1-2        1004991 non-null  int64  
 8   Crm Cd          1004991 non-null  int64  
 9   Crm Cd Desc     1004991 non-null  object 
 10  Mocodes         853372 non-null   object 
 11  Vict Age        1004991 non-null  int64  
 12  Vict Sex        860347 non-null   object 
 13  Vict Descent    860335 non-null   object 
 14  Premis Cd       1004975 non-null  float64
 15  Premis Desc     1004403 non-null  object 
 16  Weapon Used Cd  327247 non-null   fl

### View Unique values for Crimes

In [18]:
unique_desc = df["Crm Cd Desc"].dropna().unique()
len(unique_desc), unique_desc

(140,
 array(['THEFT OF IDENTITY',
        'ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT',
        'THEFT FROM MOTOR VEHICLE - GRAND ($950.01 AND OVER)',
        'THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)',
        'CRM AGNST CHLD (13 OR UNDER) (14-15 & SUSP 10 YRS OLDER)',
        'VEHICLE - STOLEN', 'BURGLARY', 'BURGLARY FROM VEHICLE',
        'THEFT PLAIN - PETTY ($950 & UNDER)',
        'INTIMATE PARTNER - SIMPLE ASSAULT', 'BATTERY - SIMPLE ASSAULT',
        'VANDALISM - MISDEAMEANOR ($399 OR UNDER)',
        'VEHICLE - ATTEMPT STOLEN',
        'VANDALISM - FELONY ($400 & OVER, ALL CHURCH VANDALISMS)',
        'ROBBERY', 'FIREARMS RESTRAINING ORDER (FIREARMS RO)',
        'BIKE - STOLEN', 'EMBEZZLEMENT, GRAND THEFT ($950.01 & OVER)',
        'CHILD ABUSE (PHYSICAL) - SIMPLE ASSAULT',
        'CRIMINAL THREATS - NO WEAPON DISPLAYED',
        'THEFT-GRAND ($950.01 & OVER)EXCPT,GUNS,FOWL,LIVESTK,PROD',
        'BATTERY WITH SEXUAL CONTACT',
        'LETTERS, LEWD  -  TELEPHONE

### Map each Crime commited to a matching Criminal Offense

In [19]:
# Keyword-based mapping rules for auto‐labeling
mapping_rules = {
    "Violent Crime": [
        "ASSAULT", "BATTERY", "HOMICIDE", "MANSLAUGHTER", "RAPE",
        "SEXUAL", "SODOMY", "ORAL COPULATION", "KIDNAPPING",
        "LYNCHING", "STALKING", "THREATS", "INTIMATE PARTNER"
    ],
    "Property Crime": [
        "THEFT", "BURGLARY", "VANDALISM", "ARSON", "SHOPLIFTING",
        "BIKE - STOLEN", "COIN MACHINE"
    ],
    "Vehicle Crime": [
        "VEHICLE", "DRIVING WITHOUT OWNER CONSENT", "DWOC"
    ],
    "Fraud / Financial Crime": [
        "FRAUD", "EMBEZZLEMENT", "COUNTERFEIT", "BUNCO",
        "CREDIT CARD", "DOCUMENT WORTHLESS", "INSURANCE"
    ],
    "Weapons / Public Safety": [
        "FIREARM", "WEAPON", "SHOTS FIRED", "BOMB", "BRANDISH"
    ],
    "Sex Crime": [
        "LEWD", "INDECENT EXPOSURE", "CHILD PORNOGRAPHY",
        "PANDERING", "PIMPING", "HUMAN TRAFFICKING"
    ],
    "Child-Related Crime": [
        "CHILD", "CONTRIBUTING", "CHILD NEGLECT"
    ],
    "Court / Restraining Order / Legal": [
        "COURT", "RESTRAINING", "CONTEMPT", "FAILURE TO APPEAR",
        "VIOLATION"
    ],
    "Public Disturbance / Disorder": [
        "DISTURBANCE", "PEACE", "TRESPASS", "DISRUPT",
        "RIOT", "DISOBEY"
    ],
    "Other Crime": []  # fallback
}

# Function to classify crimes
def classify(description: str):
    if not isinstance(description, str):
        return "Other Crime"
    desc = description.upper()
    for category, keywords in mapping_rules.items():
        for kw in keywords:
            if kw in desc:
                return category
    return "Other Crime"

# Create new class column
df["Crime_Class"] = df["Crm Cd Desc"].apply(classify)

# Save a preview
preview = df[["Crm Cd Desc", "Crime_Class"]].head(30)
preview

Crm Cd Desc     Crime_Class
0                                   THEFT OF IDENTITY  Property Crime
1      ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT   Violent Crime
2                                   THEFT OF IDENTITY  Property Crime
3   THEFT FROM MOTOR VEHICLE - GRAND ($950.01 AND ...  Property Crime
4     THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)  Property Crime
5                                   THEFT OF IDENTITY  Property Crime
6                                   THEFT OF IDENTITY  Property Crime
7   CRM AGNST CHLD (13 OR UNDER) (14-15 & SUSP 10 ...     Other Crime
8                                   THEFT OF IDENTITY  Property Crime
9                                   THEFT OF IDENTITY  Property Crime
10  CRM AGNST CHLD (13 OR UNDER) (14-15 & SUSP 10 ...     Other Crime
11                                  THEFT OF IDENTITY  Property Crime
12                                   VEHICLE - STOLEN   Vehicle Crime
13                                           BURGLARY  Property Crime
14                              BURGLARY FROM VEHICLE  Property Crime
15                 THEFT PLAIN - PETTY ($950 & UNDER)  Property Crime
16                  INTIMATE PARTNER - SIMPLE ASSAULT   Violent Crime
17                           BATTERY - SIMPLE ASSAULT   Violent Crime
18                                           BURGLARY  Property Crime
19                                           BURGLARY  Property Crime
20                 THEFT PLAIN - PETTY ($950 & UNDER)  Property Crime
21                 THEFT PLAIN - PETTY ($950 & UNDER)  Property Crime
22                 THEFT PLAIN - PETTY ($950 & UNDER)  Property Crime
23                                           BURGLARY  Property Crime
24     ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT   Violent Crime
25           VANDALISM - MISDEAMEANOR ($399 OR UNDER)  Property Crime
26                           VEHICLE - ATTEMPT STOLEN   Vehicle Crime
27  VANDALISM - FELONY ($400 & OVER, ALL CHURCH VA...  Property Crime
28                                            ROBBERY     Other Crime
29                  INTIMATE PARTNER - SIMPLE ASSAULT   Violent Crime

In [20]:
df['Crime_Class'].value_counts().reset_index(name='Count')

Crime_Class   Count
0                     Property Crime  508444
1                      Violent Crime  233487
2                      Vehicle Crime  123445
3                        Other Crime   63036
4  Court / Restraining Order / Legal   21771
5      Public Disturbance / Disorder   19977
6            Weapons / Public Safety   19431
7                          Sex Crime   11870
8                Child-Related Crime    2784
9            Fraud / Financial Crime     746

### Start a miniature testing with Decision Tree Classifier

In [21]:
# Remove Crm Cd Desc to avoid leakage
df_model = df.drop(columns=[
    "Crm Cd Desc", "Crm Cd", "Premis Cd", "Premis Desc",
    "Crm Cd 1", "Crm Cd 2", "Crm Cd 3", "Crm Cd 4"
])

# Prepare training data
X = df_model.drop(columns=["Crime_Class"])
y, _ = pd.factorize(df_model["Crime_Class"])

# Factorize object columns
X = X.apply(lambda col: pd.factorize(col)[0] if col.dtype == "object" else col)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42
)

# Train model
dt_model_1 = DecisionTreeClassifier()
dt_model_1.fit(X_train, y_train)

# Predict
y_pred_test = dt_model_1.predict(X_test)
y_pred_train = dt_model_1.predict(X_train)

# Evaluation
report_test = pd.DataFrame.from_dict(
    classification_report(y_test, y_pred_test, output_dict=True)
).transpose()

report_train = pd.DataFrame.from_dict(
    classification_report(y_train, y_pred_train, output_dict=True)
).transpose()

# Accuracy scores
train_accuracy = accuracy_score(y_train, y_pred_train)
test_accuracy = accuracy_score(y_test, y_pred_test)

print("Training Set Accuracy:", train_accuracy)
print("Testing Set Accuracy:", test_accuracy)
print("------------------------------------------------------------------------------------")
print("Training Set Report")
print("------------------------------------------------------------------------------------")
print(report_train)
print("------------------------------------------------------------------------------------")
print("Testing Set Report")
print("------------------------------------------------------------------------------------")
print(report_test)
print("------------------------------------------------------------------------------------")


Training Set Accuracy: 0.9998877032180846
Testing Set Accuracy: 0.8045161161931422
------------------------------------------------------------------------------------
Training Set Report
------------------------------------------------------------------------------------
              precision    recall  f1-score        support
0              0.999868  0.999983  0.999926  356035.000000
1              0.999982  0.999994  0.999988  163391.000000
2              0.999525  0.999729  0.999627   44202.000000
3              0.999907  0.999317  0.999612   86340.000000
4              1.000000  0.999926  0.999963   13595.000000
5              1.000000  1.000000  1.000000    8282.000000
6              1.000000  1.000000  1.000000   15207.000000
7              1.000000  1.000000  1.000000   13944.000000
8              1.000000  1.000000  1.000000     540.000000
9              1.000000  1.000000  1.000000    1957.000000
accuracy       0.999888  0.999888  0.999888       0.999888
macro avg      0.99

### Random Forest - ChiaHui

In [29]:
# RANDOM FOREST CLASSIFIER IMPLEMENTATION
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, KFold, cross_val_score
import pandas as pd

print("=== Random Forest Crime Classification ===")

# Remove Crm Cd Desc to avoid leakage
df_model = df.drop(columns=[
    "Crm Cd Desc", "Crm Cd", "Premis Cd", "Premis Desc",
    "Crm Cd 1", "Crm Cd 2", "Crm Cd 3", "Crm Cd 4"
])

# Prepare training data
X = df_model.drop(columns=["Crime_Class"])
y, class_names = pd.factorize(df_model["Crime_Class"])

# Convert datetime columns to int timestamps
for col in X.select_dtypes(include=['datetime', 'datetimetz']).columns:
    X[col] = X[col].view('int64')

# Factorize object columns
X = X.apply(lambda col: pd.factorize(col)[0] if col.dtype == "object" else col)

# Train-test split (stratify for class balance)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

# Initialize Random Forest with balanced, regularized params
rf_model = RandomForestClassifier(
    n_estimators=120,
    max_depth=22,
    min_samples_split=12,
    min_samples_leaf=6,
    max_features='sqrt',
    class_weight='balanced',
    n_jobs=-1,
    random_state=42
)

rf_model.fit(X_train, y_train)

# Predictions
y_pred_test = rf_model.predict(X_test)
y_pred_train = rf_model.predict(X_train)

# Accuracy
train_accuracy = accuracy_score(y_train, y_pred_train)
test_accuracy = accuracy_score(y_test, y_pred_test)

print("\n=== Performance ===")
print(f"Training Set Accuracy: {train_accuracy}")
print(f"Testing Set Accuracy:  {test_accuracy}")
print(f"Overfit Gap:          {train_accuracy - test_accuracy:.4f}")

# Evaluation reports
report_test = pd.DataFrame.from_dict(
    classification_report(y_test, y_pred_test, output_dict=True)
).transpose()

report_train = pd.DataFrame.from_dict(
    classification_report(y_train, y_pred_train, output_dict=True)
).transpose()

print("------------------------------------------------------------------------------------")
print("Training Set Report")
print("------------------------------------------------------------------------------------")
print(report_train)
print("------------------------------------------------------------------------------------")
print("Testing Set Report")
print("------------------------------------------------------------------------------------")
print(report_test)
print("------------------------------------------------------------------------------------")

=== Random Forest Crime Classification ===

=== Performance ===
Training Set Accuracy: 0.850693610313109
Testing Set Accuracy:  0.8027383266223989
Overfit Gap:          0.0480
------------------------------------------------------------------------------------
Training Set Report
------------------------------------------------------------------------------------
              precision    recall  f1-score        support
0              0.982384  0.815097  0.890956  355910.000000
1              0.936730  0.875778  0.905229  163441.000000
2              0.674720  0.737994  0.704940   44125.000000
3              0.823046  0.951974  0.882828   86411.000000
4              0.496538  0.980738  0.659286   13602.000000
5              0.321328  0.910579  0.475027    8309.000000
6              0.507173  0.939436  0.658722   15240.000000
7              0.523343  0.909039  0.664263   13984.000000
8              0.805556  1.000000  0.892308     522.000000
9              0.538526  0.993330  0.698413 

### Data Cleaning

Check for Duplicates

In [12]:
df = df.drop_duplicates()
df

Date Rptd                DATE OCC  TIME OCC  \
0        211507896          11/4/2021 0:00          7/11/2020 0:00       845   
1        201516622  10/21/2020 12:00:00 AM  10/18/2020 12:00:00 AM      1845   
2        240913563         10/12/2024 0:00  10/30/2020 12:00:00 AM      1240   
3        210704711  12/24/2020 12:00:00 AM  12/24/2020 12:00:00 AM      1310   
4        201418201          3/10/2020 0:00  09/29/2020 12:00:00 AM      1830   
...            ...                     ...                     ...       ...   
1004986  252104112           2/2/2025 0:00           2/2/2025 0:00       130   
1004987  250404100  02/18/2025 12:00:00 AM  02/18/2025 12:00:00 AM      1000   
1004988  251304095  01/31/2025 12:00:00 AM  01/30/2025 12:00:00 AM      1554   
1004989  251704066  01/17/2025 12:00:00 AM  01/17/2025 12:00:00 AM      1600   
1004990  251904210  03/25/2025 12:00:00 AM  03/25/2025 12:00:00 AM      1235   

         AREA    AREA NAME  Rpt Dist No  Part 1-2  \
0          15  N Hollywood         1502         2   
1          15  N Hollywood         1521         1   
2           9     Van Nuys          933         2   
3           7     Wilshire          782         1   
4          14      Pacific         1454         1   
...       ...          ...          ...       ...   
1004986    21      Topanga         2103         2   
1004987     4   Hollenbeck          479         2   
1004988    13       Newton         1372         2   
1004989    17   Devonshire         1774         2   
1004990    19      Mission         1944         2   

                                         Mocodes  Vict Age Vict Sex  \
0                                            377        31        M   
1        0416 0334 2004 1822 1414 0305 0319 0400        32        M   
2                                            377        30        M   
3                                            344        47        F   
4                            1300 0344 1606 2032        63        M   
...                                          ...       ...      ...   
1004986                                      NaN        35        M   
1004987                           1258 0553 0602        11        M   
1004988                                      NaN        16        F   
1004989                      0400 1259 1822 0356        17        M   
1004990                      0529 2024 1815 0913        35        F   

        Vict Descent  Weapon Used Cd  \
0                  H             NaN   
1                  H           200.0   
2                  W             NaN   
3                  A             NaN   
4                  H             NaN   
...              ...             ...   
1004986            X             NaN   
1004987            B             NaN   
1004988            H             NaN   
1004989            H           400.0   
1004990            H             NaN   

                                            Weapon Desc Status  Status Desc  \
0                                                   NaN     IC  Invest Cont   
1                      KNIFE WITH BLADE 6INCHES OR LESS     IC  Invest Cont   
2                                                   NaN     IC  Invest Cont   
3                                                   NaN     IC  Invest Cont   
4                                                   NaN     IC  Invest Cont   
...                                                 ...    ...          ...   
1004986                                             NaN     IC  Invest Cont   
1004987                                             NaN     IC  Invest Cont   
1004988                                             NaN     IC  Invest Cont   
1004989  STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)     IC  Invest Cont   
1004990                                             NaN     IC  Invest Cont   

                                         LOCATION Cross Street      LAT  \
0         7800    BEEMAN                       AV          NaN  34.2124   
1 

None found.

### Removing Columns 
Do not proceed without caution. In this context, the chosen columns were dropped due to them having a direct relationship to the target class, which may cause a leakage.

In [ ]:
df = df.drop(columns=["Crm Cd Desc", "Crm Cd", "Premis Cd", "Premis Desc", "Crm Cd 1", "Crm Cd 2", "Crm Cd 3", "Crm Cd 4"])
df

### Standardization

In this phase, TAKE NOTE that anything, once you hit run, results in the dataframe being permanently changed. So to rerun this segment, you MUST reload and reimport the DataFrame again.

In [ ]:
# 1. Clean DATE OCC (mixed formats)
df['DATE OCC'] = pd.to_datetime(df['DATE OCC'], format='mixed', errors='coerce')

# 2. Clean TIME OCC (force numeric → Int64 → 4-digit HHMM)
df['TIME OCC'] = pd.to_numeric(df['TIME OCC'], errors='coerce').astype('Int64')
time_str = df['TIME OCC'].astype(str).str.zfill(4)

# 3. Combine DATE OCC + TIME OCC into a single datetime
df['DateTime OCC'] = pd.to_datetime(
    df['DATE OCC'].dt.strftime('%Y-%m-%d') + ' ' + time_str,
    errors='coerce'
)

# 4. Drop the original columns used for merging
df = df.drop(columns=['DATE OCC', 'TIME OCC'])

df

### Check for NULL values

In [ ]:
df.isna().sum()

Leave as NULL, since some records do in fact not possess the given info.

In [ ]:
df = df.drop(columns=[' '])
df

Remove the Weapon Used Cd column, and change the Weapon Desc column to binary.

In [ ]:
# 1. Drop the Weapon Used Cd column (if it exists)
df = df.drop(columns=['Weapon Used Cd'], errors='ignore')

# 2. Create a binary Weapon_Present column
df['Weapon_Present'] = df['Weapon Desc'].apply(
    lambda x: 'Present' if pd.notna(x) and str(x).strip() != '' else 'Absent'
)

# 3. (Optional) Drop Weapon Desc if you want to fully remove the text info
df = df.drop(columns=['Weapon Desc'], errors='ignore')

df

### Dropping columns that provide meaningless value:

Start with obvious ones, Location of the crime will often be unique among all records, but the area/district may not. So it can be dropped.

To do so, apply a test such as Pearson Correlation

In [ ]:
plt.figure(figsize=(14, 6))
sns.heatmap(df.isna(), cbar=False, yticklabels=False)
plt.title("Missing Value Heatmap")
plt.show()

Cross Street possess a lot of missing values. This can be dropped completely if Location is already sufficient info.

As for Mo Codes, this is neglectable, since some records may only be a minor offense. So no MO Code may have been recorded.

This applies to victim sex and descent as well, since not all crimes have a victim involved, such as vandalisme.

In [ ]:
# Select only numeric columns
num_df = df.select_dtypes(include=['number'])

# Calculate correlation matrix
corr = num_df.corr(numeric_only=True)

# Turn it into a sorted report (pairwise correlations)
corr_report = (
    corr.where(~np.tril(np.ones(corr.shape)).astype(bool))  # keep upper triangle
        .stack()
        .reset_index()
)
corr_report.columns = ["Feature 1", "Feature 2", "Correlation"]

# Sort by absolute correlation strength
corr_report = corr_report.iloc[corr_report['Correlation'].abs().sort_values(ascending=False).index]

corr_report.head(20)   # view top 20 strongest relationships


### What can be derived freom this correlation table?

1. Area and Rpt Dist No are inheritelly the same info, if you know the area where the crime occured, is the specific District Number really required?
2. Many features have a correlation near 0, this implies that many features are highly independent. In  which, leave them be.
3. Latitude and Longitude do provide the coordinates of the crime, but neither predicts each other.


In [ ]:
# Make a copy of the data
df_corr = df.copy()

# Convert Crime_Class (categorical) → numeric labels
df_corr['Crime_Class_numeric'] = df_corr['Crime_Class'].astype('category').cat.codes

# Select only numeric columns
num_df = df_corr.select_dtypes(include=['number'])

# Compute correlation with the numeric-encoded target
target_corr = num_df.corr(numeric_only=True)['Crime_Class_numeric']

# Remove the target itself
target_corr = target_corr.drop(labels=['Crime_Class_numeric'])

# Turn into sorted dataframe
target_corr_report = (
    target_corr
        .abs()
        .sort_values(ascending=False)
        .rename("Correlation_with_Crime_Class")
        .to_frame()
)

target_corr_report.head(20)


Perform a mini test BEFORE dropping columns.

In [ ]:
df.info()

In [ ]:
# Remove Crm Cd Desc to avoid leakage
df_model_1 = df.copy()

# Prepare training data
X = df_model_1.drop(columns=["Crime_Class"])
y, _ = pd.factorize(df_model_1["Crime_Class"])

# Convert datetime columns to int64 timestamps
for col in X.select_dtypes(include=["datetime", "datetimetz"]).columns:
    X[col] = X[col].view('int64')

# Factorize object columns
X = X.apply(lambda col: pd.factorize(col)[0] if col.dtype == "object" else col)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42
)

# Train model
dt_model_2 = DecisionTreeClassifier()
dt_model_2.fit(X_train, y_train)

# Predict
y_pred_test = dt_model_2.predict(X_test)
y_pred_train = dt_model_2.predict(X_train)

# Evaluation
report_test = pd.DataFrame.from_dict(
    classification_report(y_test, y_pred_test, output_dict=True)
).transpose()

report_train = pd.DataFrame.from_dict(
    classification_report(y_train, y_pred_train, output_dict=True)
).transpose()

# Accuracy scores
train_accuracy = accuracy_score(y_train, y_pred_train)
test_accuracy = accuracy_score(y_test, y_pred_test)

print("Training Set Accuracy:", train_accuracy)
print("Testing Set Accuracy:", test_accuracy)
print("------------------------------------------------------------------------------------")
print("Training Set Report")
print("------------------------------------------------------------------------------------")
print(report_train)
print("------------------------------------------------------------------------------------")
print("Testing Set Report")
print("------------------------------------------------------------------------------------")
print(report_test)
print("------------------------------------------------------------------------------------")


### Let's tackle Mocodes first, these represent a given crime/offense commited. You can find the full list in the file attached.

In [ ]:
# --- Step 1: Clean & explode the MO Codes column ---
# Convert NaN to empty string
df['Mocodes'] = df['Mocodes'].fillna('')

# Split by spaces → expand into list
df['MOCODES_LIST'] = df['Mocodes'].str.strip().str.split()

# Explode (each code becomes a row)
exploded = df.explode('MOCODES_LIST')

Then, extract all unique MO code entries present.

In [ ]:
all_codes = sorted({code for sublist in df['MOCODES_LIST'] for code in sublist})
print(len(all_codes), "unique MO codes found")

Count the frequency of each MO Code in the dataset

In [ ]:
# --- Step 2: Count MO code frequencies ---
mo_counts = (
    exploded['MOCODES_LIST']
    .value_counts()
)

Select the top 100 MO Codes to keep

In [ ]:
# --- Step 3: Select the Top 100 codes ---
top_100 = set(mo_counts.head(100).index)

Filter MO list into top codes and “others”

In [ ]:
# --- Step 4: Create one-hot columns for each top code ---
for code in top_100:
    df[f"MO_{code}"] = df['MOCODES_LIST'].apply(lambda lst: code in lst)

# --- Step 5: Create the OTHERS column ---
# OTHERS = true if the row contains any MO code NOT in the top 100
df['MO_OTHERS'] = df['MOCODES_LIST'].apply(
    lambda lst: any(code not in top_100 for code in lst)
)

Filter each crime’s MO list to only keep top 100 codes, create an Others column to store everything else.

Multi-hot encode only the top 100 codes

In [ ]:
# --- Step 6: Convert booleans to integers (0/1) ---
mo_cols = [col for col in df.columns if col.startswith("MO_")]
df[mo_cols] = df[mo_cols].astype(int)

# --- Step 7: Clean up temporary column ---
df = df.drop(columns=["MOCODES_LIST"])

# --- Done ---
print(f"Created {len(mo_cols)} MO Code features (100 Top + OTHERS).")
print(mo_cols[:10])

In [ ]:
df = df.drop(columns=['Mocodes'], errors='ignore')
df

### Next, tackle the location based columns

As mentioned, since all crimes will more times than not, occur in unique, varying locations, it is best to assume that there is no association or pattern to be determined from them. They are simply too specific to be trained upon.

In [ ]:
df = df.drop(columns=['LOCATION', 'Cross Street'], errors='ignore')
df

### Next, we need to choose bewteen keeping AREA, or DISTRICT.

AREA:
- It encodes neighborhood-level crime patterns
- It’s stable and interpretable

Rpt District:
- This is a finer-grained region ID.
- Usually LAPD districts are ~1–2 square miles.

But using both AREA and Rpt Dist No creates strong multicollinearity, because:
- AREA is a parent region
- Rpt Dist No is the subregion

Which is Better?

Refer back to the correlation test, Rpt District is SLIGHTLY better than AREA, so that is what we will keep. It may contain more info than AREA, as AREA is a bit too general.

In [ ]:
df = df.drop(columns=['AREA'])

In [ ]:
df = df.drop(columns=['AREA NAME'])
df

### Finally, we need to decide what to do with LAT and LON

These are very powerful IF transformed.

Raw lat/lon are NOT useful directly because:
- models cannot interpret earth geometry
- correlation is near zero
- linear models especially fail with raw coordinates

Raw latitude and longitude values:
- have no linear meaning
- give almost zero Pearson correlation
- confuse tree models (too many splits)
- confuse linear models (not linear!)
- are extremely sensitive to tiny changes

But crimes happen in spatial hotspots:
- Downtown
- Hollywood
- South LA
- Venice
- San Fernando Valley
- Pico-Union
- Koreatown
- Westlake
- etc.

### Is there a way to make them useful via transformation?

### Clustering the LAT and LON into Bins of range values, may provide much more use.

In [ ]:
from sklearn.cluster import KMeans

coords = df[['LAT', 'LON']].dropna()

kmeans = KMeans(n_clusters=100, random_state=42)

df['Location_Cluster'] = kmeans.fit_predict(coords)

### Method 1 — KMeans Clustering

This learns 50–200 “crime regions” directly from the data.

In [ ]:
df

### Method 2 — Spatial Binning (ranges)

This converts lat/lon into a grid.

In [ ]:
df['Lat_bin'] = pd.cut(df['LAT'], bins=50, labels=False)
df['Lon_bin'] = pd.cut(df['LON'], bins=50, labels=False)


In [ ]:
df

This creates 2 features:
- Lat_bin
- Lon_bin

Which together form a 2D grid, like “Region (12, 34)”.

This is good for:
- linear models
- tree models
- giant datasets
- preserving spatial structure

### 🟢 Combining Both Is Even Better

Crime prediction systems often use:

✔ Location_Cluster (KMeans)

✔ Lat_bin + Lon_bin (50x50 grid)

This gives:
- global structure (clusters)
- local structure (grid bins)
- Without storing raw LAT/LON.

### Now, let's try building the Decision Tree Classifier again.

In [ ]:
# Remove Crm Cd Desc to avoid leakage
df_model_2 = df.copy()

# Prepare training data
X = df_model_2.drop(columns=["Crime_Class"])
y, _ = pd.factorize(df_model_2["Crime_Class"])

# Convert datetime columns to int64 timestamps
for col in X.select_dtypes(include=["datetime", "datetimetz"]).columns:
    X[col] = X[col].view('int64')

# Convert list columns to strings so they can be factorized
for col in X.columns:
    if X[col].apply(lambda x: isinstance(x, list)).any():
        X[col] = X[col].astype(str)

# Factorize object columns
X = X.apply(lambda col: pd.factorize(col)[0] if col.dtype == "object" else col)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42
)

# Train model
dt_model_3 = DecisionTreeClassifier()
dt_model_3.fit(X_train, y_train)

# Predict
y_pred = dt_model_3.predict(X_test)
y_pred_train = dt_model_3.predict(X_train)

# Evaluation
report_test = pd.DataFrame.from_dict(
    classification_report(y_test, y_pred, output_dict=True)
).transpose()

report_train = pd.DataFrame.from_dict(
    classification_report(y_train, y_pred_train, output_dict=True)
).transpose()

# Accuracy scores
train_accuracy = accuracy_score(y_train, y_pred_train)
test_accuracy = accuracy_score(y_test, y_pred)

print("Training Set Accuracy:", train_accuracy)
print("Testing Set Accuracy:", test_accuracy)
print("------------------------------------------------------------------------------------")
print("Training Set Report")
print("------------------------------------------------------------------------------------")
print(report_train)
print("------------------------------------------------------------------------------------")
print("Testing Set Report")
print("------------------------------------------------------------------------------------")
print(report_test)
print("------------------------------------------------------------------------------------")

A highly Noticeable increase.

### Build a Confusion Matrix of Errors

In [ ]:
from sklearn.metrics import confusion_matrix, classification_report
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd

# Compute confusion matrix
cm = confusion_matrix(y_test, y_pred)

# Get sorted unique class labels
labels = sorted(list(set(y_test)))

# Create labeled confusion matrix
cm_labeled = pd.DataFrame(
    cm,
    index=[f"Actual {cls}" for cls in labels],
    columns=[f"Predicted {cls}" for cls in labels]
)

plt.figure(figsize=(12, 10))
sns.heatmap(cm_labeled, annot=True, fmt='d', cmap='Blues')
plt.title("Confusion Matrix (All Classes Labeled Clearly)")
plt.show()

# Classification report
print(classification_report(y_test, y_pred))



### Evaluate Feature Importance (Decision Tree)

In [ ]:
importance = dt_model_3.feature_importances_
features = X.columns

fi = pd.DataFrame({"feature": features, "importance": importance})
fi.sort_values(by="importance", ascending=False).head(20)

In [ ]:
import shap

explainer = shap.TreeExplainer(dt_model_3, feature_perturbation="interventional")
shap_values = explainer.shap_values(
    X_test.sample(200),
    check_additivity=False
)

shap.summary_plot(shap_values, X_test.sample(200))


### Perform a 5-Folds Cross Validation to verify the results obtained.

*This may take a while...

In [ ]:
from sklearn.model_selection import cross_val_score, KFold

# Create the model (same as your main model)
cv_model = DecisionTreeClassifier()

# Define cross-validation strategy (5 folds, shuffle for randomness)
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Compute CV scores (accuracy for each fold)
cv_scores = cross_val_score(cv_model, X, y, cv=kf, scoring="accuracy")

print("Cross-Validation Accuracy Scores:", cv_scores)
print("Mean CV Accuracy:", cv_scores.mean())
print("Standard Deviation:", cv_scores.std())


We now get:
- Accuracy for each fold
- Mean CV accuracy (main metric)
- Standard deviation → shows robustness

This addresses “Evaluate performance using relevant validation techniques.”

From which, we can conclude that:
The model is extremely stable.
- The mean accuracy is 0.8745 (87.45%)
- Every fold is almost identical — differences are only 0.1% between them.
- Standard deviation is 0.00046 → very low variance

This means:
-The model generalizes well.
- The preprocessing contributes to consistent learning.
- There is no serious data leakage anymore.
- The model is not overly sensitive to how training/test data is split.

👉 This is a VERY good sign.

In terms of Training versus Testing accuracy:

- Training Accuracy: 0.999869
- Testing Accuracy: 0.872742

This is classic decision tree overfitting:

- Training accuracy is almost 100%
→ Decision Trees memorize the training data.
- Testing accuracy is ~87%
→ Still very good performance.
- The difference (~13%) shows normal overfitting for deep decision trees.

BUT IMPORTANT:
The cross-validation accuracy (87.45%) and test accuracy (87.27%) match almost perfectly.

This means:

✔ The test result is reliable

✔ The model is not unstable

✔ Overfitting is expected but not harmful because CV matches test performance.